### ISSUE TITLE
Align ktensor tutorial with python version

### ISSUE THREAD

**Issue by DeepBlockDeepak** [sandialabs/pyttb#147]
Goal: Replicate the [TTB ktensor](https://tensortoolbox.org/ktensor_doc.html) doc within pyttb/tutorial ipynb file, similar to the Tensors.ipynb file in tutorial/.

**Comment by DeepBlockDeepak** [comment]
---

# Confusion with `tendiag()` behavior


I am having an issue with converting the ttb [tendiag()](https://tensortoolbox.org/ktensor_doc.html#12:~:text=0.8131%20%20%20%200.4423%20%20%20%200.6098-,Use%20tendiag%20or%20sptendiag%20to%20convert%20a%20ktensor%20to%20a%20ttensor.,-A%20ktensor%20can), for creating a tensor with specific values along the super-diagonal. (I'm referring to "super-diagonal" as the sequence of elements that lie along the {(1,1,1...), (2,2,2,...),..,(n,n,n,...)} line of the tensor, and not [this definition](https://mathworld.wolfram.com/Superdiagonal.html) of superdiagonal. ... I hope I'm interpreting that correctly.) 

My buggy function call: 

```python
core = ttb.tendiag(X.weights, ((R,) * X.ndims)) # Create a diagonal core.
```

I substituted MATLAB's [repmat(R,1,ndims(X))](https://tensortoolbox.org/ktensor_doc.html#12:~:text=tendiag(X.lambda%2C-,repmat(R%2C1%2Cndims(X)),-)%3B%20%25%3C%2D%2D%20Create%20a) argument with `((R,) * X.ndims)` since it seems like `repmat` should be producing a `1 x ndims(X)` array filled with `R`-values. That python expression creates a 3-tuple equivalent to (2, 2, 2), which satisfies the type() for the `shape` parameter to the pyttb tendiag(). `2` is the number of weights, and there are `3` ndims, or factor_matrices, in X.

I'm expecting this to  behave by creating a 3-dimensional tensor of shape 2x2x2 with `X.weights` == (`array([5. , 0.25])`) along the diagonal... something like:

```python
array([[[5.0, 0.0],
        [0.0, 0.0]],

       [[0.0, 0.0],
        [0.0, 0.25]]])
```


## Reproducing  the error

1. Initialize a `ktensor`.
1. Set `R` as the length of `X.weights` (`X.lambda` in MATLAB), which is equal to 2 here.
1. Create a diagonal core.
```python
import pyttb as ttb
import numpy as np
np.random.seed(0)
A = np.random.rand(4,2) # Create some data.
B = np.random.rand(3,2)
C = np.random.rand(2,2)
weights = np.array([5.0, 0.25])
X = ttb.ktensor.from_data(weights, [A,B,C]) # Create the ktensor.

R = len(X.weights) # == 2

# Create a diagonal core. --> This is where the error occurs...
core = ttb.tendiag(X.weights, ((R,) * X.ndims)) # X.ndims == 3
```

## Error Message

```python
AssertionError: Invalid use of tensor setitem
```

---

**Comment by DeepBlockDeepak** [comment]
## The `sptendiag()` method does seem to work fine, and creates a `pyttb.sptensor.sptensor`

### Setup
```python
import pyttb as ttb
import numpy as np
np.random.seed(0)
A = np.random.rand(4,2) # Create some data.
B = np.random.rand(3,2)
C = np.random.rand(2,2)
weights = np.array([5.0, 0.25])
X = ttb.ktensor.from_data(weights, [A,B,C]) # Create the ktensor.

R = len(X.weights) # == 2

core = ttb.sptendiag(X.weights, ((R, ) * X.ndims)) # Sparse diagonal core.
core
```

### Output
```python
Sparse tensor of shape 2 x 2 x 2 with 2 nonzeros 
	[0, 0, 0] = 5.0
	[1, 1, 1] = 0.25
```

**Comment by DeepBlockDeepak** [comment]
# Problems I encountered in Ktensor (_in addition to the tendiag() one above_).

There are some TODO notes and potential issues in my python Ktensor notebook:

Here are some quick notes on each one I found, and which section they correspond to. In the notebook, `KTensors.ipynb`, search for `TODO` to see where these problems are flagged. 

## 1. Redundant Code Instances (Possibly just a style question)

- **Section: 'Specifying weights in a ktensor'**
- **TODO**: Suggest using the instantiation of `X` only once here, since `X` is not modified going forward until section `Subscripted assignment for a ktensor`, and I think it would save space in many of the subsequent code blocks without confusing the user.

    Current block of code is repeated often:
    ```python
    np.random.seed(0)
    A = np.random.rand(4,2) # Create some data.
    B = np.random.rand(3,2)
    C = np.random.rand(2,2)
    weights = np.array([5.0, 0.25])
    X = ttb.ktensor.from_data(weights, [A,B,C]) # Create the ktensor.
    ```

## 2. Core Type Conversion 

- **Section: 'Use tendiag or sptendiag to convert a ktensor to a ttensor'**
- **TODO**: Confirm whether it's okay to convert `core` into a tensor-type. Current implementation leads to a ValueError due to `TTENSOR` not supporting generic cores, only tensor.

    Current block of code for conversion:
    ```python
    np.random.seed(0)
    core = ttb.sptendiag(X.weights, ((R, ) * X.ndims)) # Sparse diagonal core.
    """
    --- Converting core to tensor else ValueError for Y-instancing ---
    """
    core = ttb.tensor.from_tensor_type(core) # <--- CONVERSION HERE
    Y = ttb.ttensor.from_data(core, X.factor_matrices) # Assemble the ttensor
    ```

## 3. Subscripted Assignment 

- **Section: 'Subscripted assignment for a ktensor'**
- **TODO**: Understand why subscripting with `X[2][:, [0]]` does not work, but `X.factor_matrices[2][:, [0]]` does.

    Current block of code for assignment:
    ```python
    X.factor_matrices[2][:, [0]] = np.ones((2,1)) # Change the matrix for mode 3.
    ```

## 4. `end()` Usage

- **Section: 'Use end for the last array index'**
- **TODO**: Figure out how to use `end()` in this context.

## 5. SVD Conversion 

- **Section: "Use ktensor to store the 'skinny' SVD of a matrix"**
- **TODO**: The creation of a ktensor from the SVD of `A` doesn't seem to produce a size compatible with the MATLAB version. Also, `X.double()` is not reproducing `A`.

## 6. Data Display 

- **Section: 'Displaying data'**
- **TODO**: Could not find a `datadisp` equivalent in pyttb.

I'll submit a PR shortly linking to this issue_148 so others can review and let me know their thoughts so I can clean things up.

**Comment by ntjohnson1** [comment]
On a quick pass for the list of 6 questions

1. I think it is nice for each block to be relatively stand alone. Large ranging dependencies in code and notebooks harms readability (also if later someone modifies a cell then it could break this consistent X assumption). Can you just make a function out of those lines of code called `generate_sample_ktensor()` then use that as the first line in all the boxes? Seems like that handles the duplication concern.

2. The sptensor ttensor core is tracked on [this ](https://github.com/sandialabs/pyttb/issues/61) github issue. I think adding a note to that ticket to update the notebook when the root cause is resolved and noting in the notebook that the ticket exists to solve that part of the tutorial seems reasonable.

3. This seems like a potential bug or something we should consider. Either you or Danny can open a ticket on it. Basically `X[2]` calls `__get_item__` but get item returns a [copy ](https://github.com/sandialabs/pyttb/blob/b947f6a10c8d502c333c781bc4e1aff525ded8c0/pyttb/ktensor.py#L2276) so we are writing over a copy rather than the original data. Where `X.factor_matrices[2]` is returning a reference to the underlying data so the assignment works as expected.

IIRC Danny did the ktensor implementation so he is probably better suited to review the more involved remaining questions.

**Comment by dmdunla** [comment]
Updates:
1. Done (by @DeepBlockDeepak)
2. Already identified as an issue: #61 
3. Issue created: #165
4. Already identified as an issue: #161 
5. Issue created: #167
    - Changed from using `from_factor_matrices` to  `from_data` constructor, which allows for `weights` to be specified.
6. Issue created: #166

**Comment by dmdunla** [comment]
Closed by 472c6d8f03b8116c09dd00ac5b2d544ea7003982

Continuing issues tracked via #168
